### PR TITLE
[XLA:GPU] add cuDNN flash attention support in XLA (1st PR with only SE changes)

### PR DESCRIPTION
**Pull request by Cjkkkk**
This is the 1st PR of splitting https://github.com/openxla/xla/pull/5910 with only SE changes.

* Added flash attention forward & backward graph generation.
* Added flash attention runner.

### CODE REVIEW

File: xla/stream_executor/stream.h [hunk 1]
```diff
@@ -467,7 +467,8 @@ class Stream {
       std::optional<dnn::TensorDescriptor> activation_descriptor,
       std::optional<dnn::TensorDescriptor> mask_descriptor,
       std::optional<dnn::TensorDescriptor> bias_descriptor, double scale,
-      std::optional<double> dropout_rate, std::optional<int64_t> seed) {
+      std::optional<double> dropout_rate, std::optional<int64_t> seed,
+      bool is_flash_attention) {
```
Comment (cheshire) [inline xla/stream_executor/stream.h:471]: Having such a large number of arguments isn't great, maybe we could have an options struct instead?

File: xla/stream_executor/cuda/cuda_dnn.cc [hunk 2]
```diff
@@ -3703,7 +3707,7 @@ tsl::StatusOr<cudnn_frontend::Tensor> CreateCudnnMaskTensor(
   // Create the mask output tensor
   TF_ASSIGN_OR_RETURN(
       auto mask_out_tensor,
-      CreateCudnnTensor(dims, strides, CudnnfMHAUid::VIRTUAL_ID + 300,
+      CreateCudnnTensor(dims, strides, CudnnfMHAUid::VIRTUAL_ID + 400,
```
Comment (cheshire) [inline xla/stream_executor/cuda/cuda_dnn.cc:3710]: Could you give some context for magic constants?

**Review by cheshire — CHANGES_REQUESTED** [review]

File: xla/stream_executor/cuda/cuda_dnn.cc (hunk 2 above)
Comment (Cjkkkk) [reply xla/stream_executor/cuda/cuda_dnn.cc:3710]: Every tensor in the graph needs to have a unique ID, VIRTUAL_ID is just some large number to begin with so it does not overlap with the non-virtual tensor defined in `enum CudnnfMHAUid`. Tensor ID in scale/bias/mask/softmax/dropout will start with +200, +300, +400, +500, +600 respectively.

File: xla/stream_executor/stream.h (hunk 1 above)
Comment (Cjkkkk) [reply xla/stream_executor/stream.h:471]: We have 
```struct FusedMHAOp {
  using Signature = FusedMHASignature;
  struct Config {
    FusedMHAKind kind;
    double scale;
    const MatmulTensorDescriptor& bmm1_lhs_descriptor;
    const MatmulTensorDescriptor& bmm1_rhs_descriptor;
    const MatmulTensorDescriptor& bmm2_rhs_descriptor;
    const MatmulTensorDescriptor& intermediate_bmm2_lhs_descriptor;
    const TensorDescriptor& output_descriptor;
    std::optional<TensorDescriptor> bias_descriptor;
    std::optional<TensorDescriptor> mask_descriptor;
    std::optional<TensorDescriptor> activation_descriptor;
    std::optional<double> dropout_rate;
    std::optional<int64_t> seed;
    bool is_flash_attention;
  };```
in the `lazy_op_runner.h`. We kind of passed in each member separately right now. Do we want to just accept a config in your opinion?

**Comment by Cjkkkk** [comment]
@cheshire Update a version that contains just SE changes and a few lines in runner to make things compilable. Let;s review this PR first. Will open 2 more PR for rewriter/thunk/mlir OP. Meanwhile, keep this https://github.com/openxla/xla/pull/5910 open for tracking additional bug fixes(if any) and tests. Will Keep 3 PR for view and sync any change required.

**Comment by tdanyluk** [comment]
PR rotation: @cheshire - gentle ping

**Review by cheshire — APPROVED** [review]

**Review by tdanyluk — COMMENTED** [review]
Hi @Cjkkkk,
Sorry but clang-tidy surfaced a few errors and possible improvements, so I'm adding them in a review.

File: xla/service/gpu/gpu_fused_mha_runner.cc [hunk 3]
```diff
@@ -237,29 +239,37 @@ Status RunFusedMHABackward(GpufMHABackwardParams params, se::Stream *stream,
...
-                                             params.config->d_output,
-                                             params.config->d_bmm1_lhs,
-                                             params.config->d_bmm1_rhs,
-                                             params.config->d_bmm2_rhs,
-                                             params.config->d_s,
-                                             params.config->mask,
-                                             params.config->d_bias,
-                                             dropout_rate,
-                                             seed};
+  // TODO: set is_flash_attention to real value, set it to false for now
```
[the remark this answers is not in the document]
Comment (Cjkkkk) [reply xla/service/gpu/gpu_fused_mha_runner.cc:242]: This TODO is to remind to remove this in the next 2 PRs. I added some codes to make this PR able to compile.

**Comment by Cjkkkk** [comment]
@tdanyluk @cheshire All clang-tidy issue addressed, could you take a look?

**Comment by tdanyluk** [comment]
Thanks, trying to get it submitted internally, hopefully we'll get the approval in about a day.

**Pull request closed without merging**